badge-maker/lib/xml.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * @module
+ */
+
+'use strict'
+
+function stripXmlWhitespace(xml) {
+  return xml.replace(/>\s+/g, '>').replace(/<\s+/g, '<').trim()
+}
+
+function escapeXml(s) {
+  if (typeof s === 'number') {
+    return s
+  } else if (s === undefined || typeof s !== 'string') {
+    return undefined
+  } else {
+    return s
+      .replace(/&/g, '&amp;')
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/"/g, '&quot;')
+      .replace(/'/g, '&apos;')
+  }
+}
+
+/**
+ * Representation of an XML element
+ */
+class XmlElement {
```

Comment:
This `XmlElement` class is the real meat of what I've done here. It is a slight evolution of Chris-from-the-past's suggestion in https://github.com/badges/shields/pull/5754#issuecomment-719817096 . The big advantage of this is it allows us to just construct an `XmlElement` using unescaped/unsafe data, pass it around, and then when we call `.render()` at the end, that will deal with all the escaping for us. It also decouples the string representation of the object from how we lay out the code, solving the whitespace issues.

Reply:
I'm still horribly unfamiliar with most of the package code so largely deferring to others. However, did want to ask whether we'd anticipate any additional memory load from the class-based approach, particularly given what we saw recently in production with the got experiment.

Reply:
This will be a slightly heavier implementation than pure string concatenation, but the thing I'm chasing here is developer experience rather than performance. Making it easy for us to "fall into the pit of success" when it comes to escaping has value. Every abstraction has a cost.

I think the difference we will see here will be a lot less than the overhead associated with a full featured HTTP library but it is conjecture and this code is very much on the critical path. I think to quantify it, we need to throw a real workload at it tbh. One thing I did to is A-B it against master using the benchmark script and this seems to add somewhere around 0.02 of a millisecond to a FTB render, which I am not worried about.

One variable here is that although I would like to apply this approach to the other badge styles this PR only changes the rendering of the FTB style. I don't really know what proportion of the traffic we serve is on FTB vs other styles but anecdotally I suspect it is very small so I suspect we wouldn't really start to see a big difference until we apply the approach to the flat style. I'd be surprised if we gather much useful data just from deploying this change.

Reply:
Fair enough. Since we've shifted over to Heroku it's _much_ easier to be able to canary deploy these types of changes and quickly revert back out if needed, which makes a huge difference.

This is yet another example of where having the ability to do some load/stress testing against a prod-like env would be useful, so may be worth circling back to this at some point down the road once some other related items (runtime costs/sponsorship, etc.) have concluded.